testing/e2e/tools/agents.go

```diff
@@ -0,0 +1,191 @@
+package tools
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"os"
+
+	"github.com/pkg/errors" //nolint:gomodguard //for tests
+	log "github.com/sirupsen/logrus"
+)
+
+// PolicyOutput holds the needed data to manage the output API keys
+type PolicyOutput struct {
+	// API key the Elastic Agent uses to authenticate with elasticsearch
+	APIKey string `json:"api_key"`
+
+	// ID of the API key the Elastic Agent uses to authenticate with elasticsearch
+	APIKeyID string `json:"api_key_id"`
+
+	// The policy output permissions hash
+	PermissionsHash string `json:"permissions_hash"`
+
+	// API keys to be invalidated on next agent ack
+	ToRetireAPIKeyIds []ToRetireAPIKeyIdsItems `json:"to_retire_api_key_ids,omitempty"`
+
+	// Type is the output type. Currently only Elasticsearch is supported.
+	Type string `json:"type"`
+}
+
+type ToRetireAPIKeyIdsItems struct {
+
+	// API Key identifier
+	ID string `json:"id,omitempty"`
+
+	// Date/time the API key was retired
+	RetiredAt string `json:"retired_at,omitempty"`
+}
+
+type Agent struct {
+	ID             string `json:"id"`
+	PolicyID       string `json:"policy_id"`
+	PolicyRevision int    `json:"policy_revision,omitempty"`
+	DefaultAPIKey  string `json:"default_api_key"`
+	LocalMetadata  struct {
+		Host struct {
+			Name     string `json:"name"`
+			HostName string `json:"hostname"`
+		} `json:"host"`
+		OS struct {
+			Family   string `json:"family"`
+			Full     string `json:"full"`
+			Platform string `json:"platform"`
+		} `json:"os"`
+		Elastic struct {
+			Agent struct {
+				Version  string `json:"version"`
+				Snapshot bool   `json:"snapshot"`
+			} `json:"agent"`
+		} `json:"elastic"`
+	} `json:"local_metadata"`
+	Status  string                   `json:"status"`
+	Outputs map[string]*PolicyOutput `json:"outputs,omitempty"`
+}
+
+// GetAgentByHostnameFromList get an agent by the local_metadata.host.name property, reading from the agents list
+func (c *Client) GetAgentByHostnameFromList(hostname string) (Agent, error) {
+
+	agents, err := c.ListAgents()
+	if err != nil {
+		return Agent{}, err
+	}
+
+	for _, agent := range agents {
+		agentHostname := agent.LocalMetadata.Host.Name
+		// a hostname has an agentID by status
+		if agentHostname == hostname {
+			log.WithFields(log.Fields{
+				"agent": agent,
+			}).Trace("Agent found")
+			return agent, nil
+		}
+	}
+
+	return Agent{}, nil
+}
+
+// ListAgents returns the list of agents enrolled with Fleet.
+func (c *Client) ListAgents() ([]Agent, error) {
+	statusCode, respBody, err := c.get(context.Background(), fmt.Sprintf("%s/agents", "api/fleet"))
```

Comment:
This request needs a timeout. The best way to do this is to make it take a `context.Context` argument and then the caller can pass a context created with [context.WithTimeout()](https://pkg.go.dev/context#WithTimeout)

Reply:
Arguable you should retry a few times but we'll see if this causes test flakiness in reality.

Reply:
The retries are probably handled by assert.Eventually use elsewhere which would make sense.

Reply:
Added context to all http requests